Fix loop detection for associations with fetch="join" mapping

deAtog opened the pull request:
Fixes  #2201

fredericDelaporte commented:
It is usually more efficient to setup your development for running the tests itself, see [contributing (Setting up for development)](https://github.com/nhibernate/nhibernate-core/blob/master/CONTRIBUTING.md#setting-up-for-development).

deAtog commented:
@fredericDelaporte I have a test/debug environment but running the tests locally was taking far more time than the automated build here. I was also getting local failures where there were none on the ci system.

I'm still debating what to do here.. disabling the circular join checks causes undoubtedly causes issues when the max_fetch_depth is unset and in some cases even with it set it still causes problems.

deAtog commented:
@fredericDelaporte This should address the issue identified in issue #2201 and provide a mechanism for users to disable the loop detection which results in an 1+N case when the entity graph does not have a loop, but the query does. It should be safe for users to disable the loop detection whenever they do not have implicit loops in their entity graph or when they have explicitly set max_fetch_depth.

bahusoid reviewed (changes requested):
I will do a proper review when I have time but it looks pretty good at quick glance. Some notes:
You shouldn't delete protected members (`IsJoinable`) - it's a breaking change. Mark it as obsolete instead. And better to get rid of formatting changes in unrelated places.

You disabled in tests loop detection but reported issue is actually related to old loop detection logic. So reported issue is actually not tested. Please add tests to verify that proper entities are initialized when loop detection is enabled.

Use additionally `NHibernateUtil.IsInitialized` in tests to verify that entity is actually initialized. Something like:
```C#
Assert.That(p.Parent.Details, Is.Not.Null);
Assert.That(NHibernateUtil.IsInitialized(p.Parent.Details));
```

deAtog commented:
@bahusoid With a one-to-one relationship both the old and new behavior always initialize the one-to-one, but they cause additional EntityFetch's to occur.

The changes to the JoinWalker are well tested by every other test in the suite.  A test that verifies the old behavior vs the new behavior would be very similar to those that failed after modifying it. Those tests only passed due to the side affects of the loop detection and depth-first nature of the old JoinWalker. Those tests did the following:
1. Query for a Contract
   a. Contract has fetch="join" set for Subcontracts and Parties
   b. Subcontracts is of type Contract and is initialized before Parties. As a result, Contract.Subcontract.Parties was was being initialized but Contract.Parties was not.
2. Deleted parties for the Contract
3. Deleted the Contract
    a. Deleting a Contract cascade deletes Subcontracts, Parties, and Infos.

After the changes to the JoinWalker, Contract.Subcontract.Parties is not initialized but Contract.Parties is. Attempting to delete the Contract after deleting all Parties threw an exception because the Contract could not cascade delete the Parties that were associated with it. If these tests had originally asserted that Contracts.Parties was initialized they would have failed. 

Swapping steps 1 and 2 above, ensures that the loaded Contract does not have any Parties associated with it and the deletion of the Contract succeeds without issue.

deAtog commented:
@bahusoid I'm going to re-order/combine these commits to make it easier to review. I'll let you know when it is ready again. The recent commits revert some of the original modifications to JoinWalker to avoid using the queue for anything but required joins. This eliminates the overhead involved in the simplest of cases where no joins are present.

hazzik commented:
@bahusoid  could you please take another look? It seems the changes you've requested are implemented now.

bahusoid commented:
Pretty busy IRL right now - no time for proper reviews (maybe in June). But at really quick glance I still see no tests to verify that #2201 is actually fixed. Also it seems `currentDetph` calculation is broken for entity joins - so we also need test that verifies proper eager child loading for entity join when max_fetch_depth set to 1

hazzik commented on src/NHibernate/Loader/JoinWalker.cs:
```diff
@@ -299,26 +312,48 @@ private void AddAssociation(OuterJoinableAssociation association)
+					case EntityQueueEntry eqe:
+						WalkEntityTree(eqe.Persister, eqe.Alias, eqe.Path);
+						break;
+					default:
```
`default` is never reached here. The `QueueEntry` is always either `CollectionQueueEntry` or `EntityQueueEntry`.

hazzik replied:
Oh, I see... you enqueue `null` there.

bahusoid commented on src/NHibernate.Test/NHSpecificTest/GH2201/CircularReferenceFetchDepthFixture.cs:
```diff
@@ -0,0 +1,65 @@
+		[Test]
+		public void QueryOver()
+		{
+			using(var logSpy = new SqlLogSpy())
```
Yeap.. Removed

hazzik commented on src/NHibernate/Loader/JoinWalker.cs:
```diff
@@ -1202,5 +1253,53 @@ protected static string GetSelectFragment(OuterJoinableAssociation join, string
 			return join.GetSelectFragment(entitySuffix, collectionSuffix, next);
 		}
+
+		protected abstract class QueueEntry
```
This should be just interface

hazzik commented on src/NHibernate/Loader/JoinWalker.cs:
```diff
@@ -1202,5 +1253,53 @@ protected static string GetSelectFragment(OuterJoinableAssociation join, string
+			public abstract void Walk(JoinWalker walker);
+		}
+
+		protected abstract class QueueEntry<T> : QueueEntry where T : IJoinable
```
There's no usage for this class.

hazzik commented on src/NHibernate/Loader/JoinWalker.cs:
```diff
@@ -1202,5 +1253,53 @@ protected static string GetSelectFragment(OuterJoinableAssociation join, string
+			public T Persister { get; set; }
+		}
+
+		protected abstract class EntityQueueEntry<T> : QueueEntry<T> where T : IJoinable
```
There's no usage for this class.

hazzik commented on src/NHibernate/Loader/JoinWalker.cs:
```diff
@@ -439,31 +461,46 @@ internal OuterJoinableAssociation InitAssociation(OuterJoinableAssociation assoc
 		/// </summary>
-		protected virtual void WalkEntityTree(IOuterJoinLoadable persister, string alias, string path, int currentDepth)
+		/// Since 5.4
+		protected virtual void WalkEntityTree(IOuterJoinLoadable persister, string alias, string path)
```
Is this a new method?

bahusoid replied:
New overload without `currentDepth`. Old one is `Obsolete` now.. `Since...` comment need to be moved to old one...
